### PR TITLE
externals: add overarching USE_SYSTEM_LIBS variable

### PR DESCRIPTION
**Pull request by Castor215**
I was going to do this when I had finished making the USE_SYSTEM_* switches, but looking at the current CMakeLists.txt I imagine some people's CMake invocations are getting a little ridiculous. Anyway, this add a sole USE_SYSTEM_LIBS option that will set all other switches to ON unless specifically disabled with DISABLE_SYSTEM_*.

Making this a draft because it's still a little rough around the edges. It also needs more testing with multiple invocations of CMake with different DISABLE switches.

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/7078)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -105,6 +105,8 @@ option(USE_SYSTEM_FDK_AAC_HEADERS "Use the system fdk-aac headers (instead of th
 option(USE_SYSTEM_FFMPEG_HEADERS "Use the system FFmpeg headers (instead of the bundled one)" OFF)
 option(USE_SYSTEM_GLSLANG "Use the system glslang and SPIR-V libraries (instead of the bundled ones)" OFF)
```
Comment (Steveice10) [inline CMakeLists.txt:106]: If we're setting up a new module for handling the system library stuff can we move all these options there as well?

**Comment by Steveice10** [comment]
> Okay, I noticed while testing this that Citra now defaults to a debug build. I could have _sworn_ it was defaulting to Release a few days ago. At first I thought it was something in this PR, but I checked out the master branch and even cloned upstream to a different directory and still the same. Is this something my end? If the answer is "no" then this is ready for review. Just thought I'd check.

I think it might be on your end, at least I'm not seeing it happen.

File: externals/cmake-modules/CitraHandleSystemLibs.cmake [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    foreach(CURRENT_LIB IN LISTS LIB_VAR_LIST)
+        if(NOT DISABLE_SYSTEM_${CURRENT_LIB})
+            set(USE_SYSTEM_${CURRENT_LIB} ON CACHE BOOL "Using system ${CURRENT_LIB}" FORCE)
+        else()
+            # Explicitly disable this in case of multiple CMake invocations
+            set(USE_SYSTEM_${CURRENT_LIB} OFF CACHE BOOL "Using system ${CURRENT_LIB}" FORCE)
+        endif()
+    endforeach()
+endif()
```
Comment (FearlessTobi) [inline externals/cmake-modules/CitraHandleSystemLibs.cmake:78]: needs a newline here

**Comment by Castor215** [comment]
The defaulting to debug thing is definitely caused by the submodules in my fork. If I reset to #7075 and update the submodules it works fine, but any further commits and it defaults to debug. I think it's cryptopp that's doing it. Relevant cmake output:
#7075 
```
=> Project : cryptopp v8.7.0 (15798ac9c261)
-- Crypto++ from user-specified location at: $HOME/citra/externals/cryptopp
=> Module : cryptopp
-- [cryptopp] CMake version 3.27.7
-- [cryptopp] System Linux
-- [cryptopp] Processor x86_64
-- [cryptopp] CMAKE_HOST_SYSTEM_PROCESSOR : x86_64
-- [cryptopp]      CMAKE_SYSTEM_PROCESSOR : x86_64
-- [cryptopp] Target architecture detected as: x86_64 -> CRYPTOPP_AMD64
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSE2
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSE2 - Success
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSE3
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSE3 - Success
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSSE3
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSSE3 - Success
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSE41
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSE41 - Success
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSE42
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSE42 - Success
-- [cryptopp] Performing Test CRYPTOPP_HAVE_CLMUL
-- [cryptopp] Performing Test CRYPTOPP_HAVE_CLMUL - Success
-- [cryptopp] Performing Test CRYPTOPP_HAVE_AESNI
-- [cryptopp] Performing Test CRYPTOPP_HAVE_AESNI - Success
-- [cryptopp] Performing Test CRYPTOPP_HAVE_AVX
-- [cryptopp] Performing Test CRYPTOPP_HAVE_AVX - Success
-- [cryptopp] Performing Test CRYPTOPP_HAVE_AVX2
-- [cryptopp] Performing Test CRYPTOPP_HAVE_AVX2 - Success
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SHANI
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SHANI - Success
-- [cryptopp] Platform: x86_64
-- [cryptopp] Compiler definitions:  CRYPTOPP_DATA_DIR="$HOME/citra/externals/cryptopp"
-- [cryptopp] Compiler options:  -fno-devirtualize 
-- [cryptopp] Build type: 
-- Version: 10.1.1
-- Build type: 
```
#6942 
```
=> Project : cryptopp-cmake v8.9.0
-- Setting build type to 'Debug' as none was specified.
-- Found Git: /usr/bin/git (found version "2.42.0") 
-- Crypto++ from user-specified location at: $HOME/citra/externals/cryptopp
-- Using branch HEAD for tests
=> Module : cryptopp
-- [cryptopp] CMake version 3.27.7
-- [cryptopp] System Linux
-- [cryptopp] Processor x86_64
-- [cryptopp] CMAKE_HOST_SYSTEM_PROCESSOR : x86_64
-- [cryptopp]      CMAKE_SYSTEM_PROCESSOR : x86_64
-- [cryptopp] Target architecture detected as: x86_64 -> CRYPTOPP_AMD64
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSE2
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSE2 - Success
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSE3
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSE3 - Success
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSSE3
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSSE3 - Success
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSE41
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSE41 - Success
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSE42
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SSE42 - Success
-- [cryptopp] Performing Test CRYPTOPP_HAVE_CLMUL
-- [cryptopp] Performing Test CRYPTOPP_HAVE_CLMUL - Success
-- [cryptopp] Performing Test CRYPTOPP_HAVE_AESNI
-- [cryptopp] Performing Test CRYPTOPP_HAVE_AESNI - Success
-- [cryptopp] Performing Test CRYPTOPP_HAVE_AVX
-- [cryptopp] Performing Test CRYPTOPP_HAVE_AVX - Success
-- [cryptopp] Performing Test CRYPTOPP_HAVE_AVX2
-- [cryptopp] Performing Test CRYPTOPP_HAVE_AVX2 - Success
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SHANI
-- [cryptopp] Performing Test CRYPTOPP_HAVE_SHANI - Success
-- [cryptopp] Platform: x86_64
-- [cryptopp] Compiler definitions:  
-- [cryptopp] Compiler options:  -fno-devirtualize 
-- [cryptopp] Build type: Debug
-- Version: 10.1.1
-- Build type: Debug
```
I don't know how to fix this on my fork.

**Comment by Steveice10** [comment]
That happens within the context of cryptopp and should only affect cryptopp. But thanks for pointing that out, will take a look as cryptopp should not be in Debug mode.

**Comment by Castor215** [comment]
> That happens within the context of cryptopp and should only affect cryptopp. But thanks for pointing that out, will take a look as cryptopp should not be in Debug mode.

The problem seems to be that, rather than passively checking the CMAKE_BUILD_TYPE and setting a local varable like Citra does, cryptopp actively sets it. The relevant part of `externals/cryptopp-cmake/CMakeLists.txt (line 202)`:
```
if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
    message(
        STATUS
        "Setting build type to '${CRYPTOPP_DEFAULT_BUILD_TYPE}' as none was specified."
    )

    set(CMAKE_BUILD_TYPE
        "${CRYPTOPP_DEFAULT_BUILD_TYPE}"
        CACHE STRING
        "Choose the type of build."
        FORCE
    )
endif()
```
I don't think the top-level `CmakeLists.txt` uses CMAKE_BUILD_TYPE beyond setting the IS_DEBUG_BUILD variable. But the section I listed above clearly affects something as commenting it out causes Citra to default to Release, as it shoud. For what it's worth, [this](https://github.com/abdes/cryptopp-cmake/commit/7ff393d82eae9dd59a8feee9a6c2443209a3280e) is the relevant commit in cryptocpp-cmake.

Speaking of which, is nobody else getting this? Not even on a completely fresh build with updated submodules? I am mystified that this is only affecting me.

**Comment by Steveice10** [comment]
Yes, but changing those variables only impacts the context they were changed in and below, i.e. only CryptoPP.

Anyway I have a PR out now to fix that up.

**Comment by Steveice10** [comment]
If you rebase on master and clear your CMake cache your build type issues should hopefully be resolved.

**Comment by Castor215** [comment]
> If you rebase on master and clear your CMake cache your build type issues should hopefully be resolved.

Funny enough I just this second tested it and was about to tell you it worked, so thank you for doing that.

I have now tested this PR with pretty much every combination and I think it's ready for review now.

File: CMakeLists.txt [hunk 3]
```diff
@@ -111,6 +111,8 @@ option(USE_SYSTEM_FDK_AAC_HEADERS "Use the system fdk-aac headers (instead of th
...
```
Comment (Steveice10) [inline CMakeLists.txt:112]: Since these are in the system libs module now shouldn't they be removed from here?

Comment (Castor215) [reply CMakeLists.txt]: Oops not sure how those made it back in. Sorry.

**Review by Steveice10 — APPROVED** [review]

**Review by FearlessTobi — APPROVED** [review]

**Pull request merged**